### PR TITLE
[Pytorch Profiler] Add support for adding module hierarchy to

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62419
* #62418
* #62417
* #62228
* #62191
* __->__ #61792

KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)

Screen Shot 2021-07-16 at 4 59 05 PM

### CODE REVIEW

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

File: torch/autograd/profiler.py [hunk 1]
```diff
@@ -358,6 +358,10 @@ class profile(object):
 
         with_stack (bool, optional): record source information (file and line number) for the ops.
 
+        with_module_hierarchy (bool): record module hierarchy (including function names)
+            corresponding to the callstack of the op. e.g. If A's forward call's B's forward
+            and it has aten::add op then aten::add's module hierarchy is A.B
```
Comment (raziel) [inline torch/autograd/profiler.py:363]:
```suggestion
            corresponding to the callstack of the op. e.g. If module A's forward calls module B's forward
            which contains an aten::add op, then aten::add's module hierarchy is A.B
```

File: torch/csrc/autograd/profiler_kineto.h [hunk 2]
```diff
@@ -203,6 +217,7 @@ struct TORCH_API KinetoEvent {
   uint8_t activity_type_ = 0;
   c10::optional<std::vector<std::vector<int64_t>>> shapes_;
   c10::optional<std::vector<std::string>> stack_;
+  c10::optional<std::string> module_hierarchy_;
```
Comment (raziel) [inline torch/csrc/autograd/profiler_kineto.h:220]: Do you see any future need to parse this hierarchy that calls out for having the hierarchy broken into its different levels? Basically just wondering if this should be a vector as well.
Just wondering...

File: torch/csrc/jit/runtime/interpreter.cpp [hunk 3]
```diff
@@ -777,6 +777,79 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
   }
 
  public:
+  std::string moduleHierarchy() const {
+    std::string module_hierarchy("TOP");
+    std::vector<StackEntry> entries;
```
Comment (raziel) [inline torch/csrc/jit/runtime/interpreter.cpp:782]: entries is unused

File: torch/csrc/jit/runtime/interpreter.cpp [hunk 4]
```diff
@@ -777,6 +777,79 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
...
+      std::string g_self_type;
+      if (g && g->inputs().size() > 0) {
+        const auto& g_self_type_ptr =
+            g->inputs()[0]->type()->cast<c10::ClassType>();
+        if (g_self_type_ptr) {
+          g_self_type = g_self_type_ptr->name()->qualifiedName();
+          g_self_type = g_self_type.substr(g_self_type.find_last_of('.') + 1);
+        }
+      }
+      module_hierarchy += "(" + g_self_type + ")::" + fn_name;
```
Comment (raziel) [inline torch/csrc/jit/runtime/interpreter.cpp:799]: not be a bad suggestion to use 'operator+=' or 'string::append()'

File: torch/profiler/profiler.py [hunk 5]
```diff
@@ -116,6 +116,9 @@ class profile(object):
         with_stack (bool): record source information (file and line number) for the ops.
         with_flops (bool): use formula to estimate the FLOPS of specific operators
             (matrix multiplication and 2D convolution).
+        with_module_hierarchy (bool): record module hierarchy (including function names)
+            corresponding to the callstack of the op. e.g. If A's forward call's B's forward
+            and it has aten::add op then aten::add's module hierarchy is A.B
```
Comment (raziel) [inline torch/profiler/profiler.py:121]:
```suggestion
            corresponding to the callstack of the op. e.g. If module A's forward calls module B's forward
            which contains an aten::add op, then aten::add's module hierarchy is A.B
```

File: torch/csrc/autograd/profiler_kineto.h (hunk 2 above)
Comment (kimishpatel) [reply torch/csrc/autograd/profiler_kineto.h:220]: @raziel probably yes. I can imagine that we build a tool integrated with the profiler that provides hierarchical break down of latency, but that would require some post processing anyway (such as modules marked with "SELF" are the same module that was seen a level up in hierarchy).

File: torch/csrc/jit/runtime/interpreter.cpp (hunk 4 above)
Comment (kimishpatel) [reply torch/csrc/jit/runtime/interpreter.cpp:799]: Any particular reason as to why one vs. the other? Is it to avoid intermediate string construction?

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

File: torch/csrc/jit/runtime/interpreter.cpp (hunk 4 above)
Comment (raziel) [reply torch/csrc/jit/runtime/interpreter.cpp]: yep that's right, which is particularly relevant in the interpreter code.

**Review by raziel — COMMENTED** [review]
Looks good to me but definitely the others should review and likely have relevant comments.

File: torch/csrc/autograd/profiler_kineto.cpp [hunk 6]
```diff
@@ -268,6 +274,10 @@ void pushProfilingCallbacks() {
           }
           ctx_ptr->stack = callstackStr(cs);
         }
+        if (config.with_module_hierarchy &&
+            fn.scope() != at::RecordScope::BACKWARD_FUNCTION) {
+          ctx_ptr->module_hierarchy = jit::currentModuleHierarchy();
```
Comment (gdankel) [inline torch/csrc/autograd/profiler_kineto.cpp:279]: Can the main part of the work actually be moved to finalizeCPUTrace()?

Comment (gdankel) [reply torch/csrc/autograd/profiler_kineto.cpp:279]: Why the check for backward function here and above?

Comment (kimishpatel) [reply torch/csrc/autograd/profiler_kineto.cpp:279]: > Why the check for backward function here and above?

Good question. For now I just follow other pattern for stacktrace. I am not 100% sure if same information is available for backward functions.

Comment (kimishpatel) [reply torch/csrc/autograd/profiler_kineto.cpp:279]: > Can the main part of the work actually be moved to finalizeCPUTrace()?

Let me check. It has same constraints as `jit::currentStackTrace`. By the time we go to finalizeCPUTrace scope of JIT is gone and we wont get any information.

For the purpose of reducing overhead, we may have to do it differently, but I havent yet thought about it.

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

File: torch/csrc/autograd/init.cpp [hunk 7]
```diff
@@ -73,7 +73,13 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
       .value("CUDA", ActivityType::CUDA);
 
   py::class_<ProfilerConfig>(m, "ProfilerConfig")
-      .def(py::init<ProfilerState, bool, bool, bool, bool>());
+      .def(py::init<ProfilerState,
+          bool, /* record_input_shapes */
+          bool, /* profile_memory */
+          bool, /* with_stac k*/
```
Comment (raziel) [inline torch/csrc/autograd/init.cpp:79]:
```suggestion
          bool, /* with_stack*/
```

File: torch/csrc/jit/runtime/interpreter.cpp [hunk 8]
```diff
@@ -776,6 +777,95 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
   }
 
  public:
+  // One way to avoid overhead of forming string would be to return
```
Comment (raziel) [inline torch/csrc/jit/runtime/interpreter.cpp:780]: Currently this returns the hierarchy as a vector of module definitions as mangled <type>::<function_name> (plus some special keys for things like top modules, etc.).

I suggest documenting this as well (i.e. what this function actually returns).

**Review by raziel — APPROVED** [review]

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

**Pull request closed without merging**